torch/distributed/tensor/parallel/recompute_wrapper.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import contextlib
+from functools import partial
+from typing import Any, Optional
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor, Replicate, Shard
+from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
+    CheckpointImpl,
+    CheckpointWrapper,
+)
+
+__all__ = [
+    "tp_checkpoint_wrapper",
+]
+
+
+class TPCheckpointWrapper(CheckpointWrapper):
```

Comment:
I'm curious why you subclass from CheckpointWrapper here?
Why is this not an independent feature like AutoReshard() and then the caller is expected to do `AutoReshard(CheckpointWrapper(mod))` ?

That would allow you to separate the discussion about what this code does from the checkpoint code?

Reply:
because we want to this wrapper to be compatible with 2D parallel, so that it can work with a module wrapped with FSDP.

Reply:
i dunno if you answered alban's question (technically) - couldn't you use Reshard(Checkpoint()) together with FSDP? I guess the question was about whether it needs to be a monolithic thing to work correctly, or if it can be 2 things that can compose

Reply:
Is the Reshard here a self-defined autograd function or?

Reply:
Reshard would be exactly the implementation you have here minus the call into the checkpoint Module.
So they can compose exactly the same way but you don't have to use Checkpoint when you use Reshard.

Reply:
Oh, I see your point, maybe I can make it just a subclass of `nn.Module`. Thanks!

Reply:
That sounds great! Thanks!

Reply:
Addressed your feedback, LMK if you have other comments.